rfcs/026-work-enricher/README.md

```diff
@@ -0,0 +1,160 @@
+
+## Graph Store
+
+There has been some investigation into the use of graph databases such as Neo4j and AWS Neptune for storage and analysis of concepts in the pipeline. Using a specialised graph store might be a good general solution for relationships within our data.
+
+Neo4j for example would enable us, if desired, to store heterogeneous data types and relationships within a single graph:
+
+```
+[ Work A ] ── (matches with) ── [ Work C ]
+    │
+(parent of)
+    │
+[ Work B ]
+    │
+(relates to)
+    │
+    ├── [ Concept A ]
+    └── [ Concept B ]
+```
+
+Here square brackets indicate entities and parentheses indicate (directed) relationships.
+
+Having a single graph representation would enable richer queries and analytics of our data, and would likely be especially useful for machine learning work. However this may introduce complexity by breaking assumptions about data flow in the pipeline, such as if there were multiple stages which had the ability to write to the graph. This should therefore should be considered with caution. Whether as part of a single graph representation or a more specialised store solely for inter work relationships in the context of archives, using such a database to store relationships would aid implementation of the denormalisation stage, especially with the traversal and emitting of related works.
+
+Rather than storing the full work within the graph database, if only the IDs and relationships are stored this allows us more flexibility about where the graph analysis is performed in the pipeline and prevent data inconsistency errors in cases where the analysis was performed on pre-merged works. The current matcher database also stores solely the IDs, with the merger using these IDs to fetch the works themselves from the recorder VHS.
+
+There are a number of graph and non graph stores we could use for storing relationships:
+
+* [Neo4j](https://neo4j.com) is the most well known graph store and probably the most advanced. The version 4 release from this year introduces sharding and thus provides horizontal scalability. We could run this hosted with [Neo4j Aura](https://neo4j.com/aura) or deploy to AWS ourselves. Scala supports seems OK, with a [DSL library](https://github.com/manishkkatoch/scala-cypher-dsl) which seems semi-official but not exactly widely used.
+
+* [AWS Neptune](https://aws.amazon.com/neptune) is a fully managed graph database solution from Amazon. Querying is done via Gremlin or SPARQL.
```

Comment:
I think I'm inclined towards Neptune just because of how much of a pain DB management can be... Neo4j docs have some quite opaque stuff that implies scalability is a feature of the Enterprise Edition (🙄 ) only - is that true?

Reply:
weco-bot.... have you become sentient?

Reply:
but I agree about Aura anyway

Reply:
general point about db management is fair, but afaict scalability is one of aura's headline features - would be very surprised if it was an add-on

Reply:
@harrisonpim assuming you mean neo4j rather than aura, I agree but got that idea from [here](https://neo4j.com/business-edge/whats-in-neo4j-enterprise-edition/)

> Scaling Neo4j horizontally over multiple machines is one of the greatest benefits of using the Enterprise Edition

Perhaps that's what v4 brings to the community edition?

Reply:
i was referring to aura as a one of [three neo4j editions](https://neo4j.com/subscriptions/) (community, enterprise, aura). from what i can see, enterprise and aura both have scaling as a prominent feature

Reply:
Ah right OK - so I think that mean that we couldn't run horizontally scalable Neo4j ourselves

Reply:
FWIW, as part of the original investigation for the matcher merger way back when I di look into Neptune or whatever it was called back then. Gremlin is horrible and everyone knows my opinions about SPARQL, so I would lean towards neo4j moslty because it's well established as a graph database and the language makes a lot more sense. Even if Aura is new and it definitely will have teething problems, I would expect them to be less long-lasting than having to deal with a horrible query language forever